h3ronpy/src/op/localij.rs

```diff
@@ -1,78 +1,84 @@
 use h3arrow::array::CellIndexArray;
 use h3arrow::h3o::CellIndex;
 use pyo3::exceptions::PyValueError;
 use pyo3::prelude::PyAnyMethods;
-use pyo3::{pyfunction, Bound, PyAny, PyObject, PyResult, Python, ToPyObject};
+use pyo3::{pyfunction, Bound, PyAny, PyObject, PyResult, Python};
+use pyo3_arrow::error::PyArrowResult;
+use pyo3_arrow::PyTable;
 use std::iter::repeat;
+use std::sync::Arc;
 
 #[pyfunction]
 #[pyo3(signature = (cellarray, anchor, set_failing_to_invalid = false))]
 pub(crate) fn cells_to_localij(
-    cellarray: &Bound<PyAny>,
+    py: Python<'_>,
+    cellarray: PyConcatedArray,
     anchor: &Bound<PyAny>,
     set_failing_to_invalid: bool,
 ) -> PyResult<PyObject> {
-    let cellindexarray = pyarray_to_cellindexarray(cellarray)?;
+    let cellindexarray = cellarray.into_cellindexarray()?;
     let anchorarray = get_anchor_array(anchor, cellindexarray.len())?;
 
     let localij_arrays = cellindexarray
         .to_local_ij_array(anchorarray, set_failing_to_invalid)
         .into_pyresult()?;
 
-    with_pyarrow(|py, pyarrow| {
-        let arrays = [
-            localij_arrays.i.into_data().to_pyarrow(py)?,
-            localij_arrays.j.into_data().to_pyarrow(py)?,
-            localij_arrays
-                .anchors
-                .primitive_array()
-                .into_data()
-                .to_pyarrow(py)?,
-        ];
-        let table = pyarrow
-            .getattr("Table")?
-            .call_method1("from_arrays", (arrays, ["i", "j", "anchor"]))?;
-        Ok(table.to_object(py))
-    })
+    let outarrays: Vec<ArrayRef> = vec![
+        Arc::new(localij_arrays.i),
+        Arc::new(localij_arrays.j),
+        Arc::new(PrimitiveArray::from(localij_arrays.anchors)),
+    ];
+
+    let schema = Arc::new(Schema::new(vec![
+        Field::new("i", outarrays[0].data_type().clone(), true),
+        Field::new("j", outarrays[1].data_type().clone(), true),
+        Field::new("anchor", outarrays[2].data_type().clone(), true),
+    ]));
+
+    let rb = RecordBatch::try_new(schema.clone(), outarrays).into_pyresult()?;
+
+    PyTable::try_new(vec![rb], schema)?.to_arro3(py)
 }
 
 #[pyfunction]
 #[pyo3(signature = (anchor, i_array, j_array, set_failing_to_invalid = false))]
 pub(crate) fn localij_to_cells(
+    py: Python<'_>,
     anchor: &Bound<PyAny>,
-    i_array: &Bound<PyAny>,
-    j_array: &Bound<PyAny>,
+    i_array: PyConcatedArray,
+    j_array: PyConcatedArray,
     set_failing_to_invalid: bool,
-) -> PyResult<PyObject> {
-    let i_array = pyarray_to_native::<Int32Array>(i_array)?;
-    let j_array = pyarray_to_native::<Int32Array>(j_array)?;
+) -> PyArrowResult<PyObject> {
+    let i_array = i_array.into_int32array()?;
+    let j_array = j_array.into_int32array()?;
     let anchorarray = get_anchor_array(anchor, i_array.len())?;
 
     let localij_arrays = LocalIJArrays::try_new(anchorarray, i_array, j_array).into_pyresult()?;
 
-    let cellarray = if set_failing_to_invalid {
```

Comment:
FYI not holding the `py: Python` token does **not** mean that it will automatically release the GIL. You always need to manually release the GIL with `py.allow_threads`

Reply:
Yes. I misunderstood that and already used allow_threads in many places.

Reply:
FWIW I also misunderstood that at first and assumed that not holding `py` meant you automatically would release the GIL. It would be really cool if it did work that way haha